File: server/src/test/java/com/todolist/server/sync/MergePolicyTest.java
package com.todolist.server.sync;

import com.todolist.server.sync.MergePolicy.Verdict;
import org.junit.jupiter.api.Test;

import java.time.Instant;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.assertEquals;

class MergePolicyTest {

    private static final Instant T = Instant.parse("2026-09-07T00:00:00Z");
    private static final UUID DEV_A = UUID.fromString("00000000-0000-4000-8000-00000000000a");
    private static final UUID DEV_B = UUID.fromString("00000000-0000-4000-8000-00000000000b");

    @Test
    void newerClientWins() {
        assertEquals(Verdict.ACCEPT, MergePolicy.decide(T.plusSeconds(5), DEV_A,
                T, DEV_B));
    }

    @Test
    void olderClientLoses() {
        assertEquals(Verdict.CONFLICT, MergePolicy.decide(T.minusSeconds(5), DEV_A,
                T, DEV_B));
    }

    @Test
    void equalTimeTieBreakByDeviceId() {
        // DEV_B > DEV_A（字典序）→ 由 B 提交的版本胜出
        assertEquals(Verdict.CONFLICT, MergePolicy.decide(T, DEV_A, T, DEV_B));
        assertEquals(Verdict.ACCEPT, MergePolicy.decide(T, DEV_B, T, DEV_A));
    }

    @Test
    void equalTimeSameDeviceSelfAccept() {
        assertEquals(Verdict.ACCEPT, MergePolicy.decide(T, DEV_A, T, DEV_A));
    }

    @Test
    void equalTimeNullServerDeviceAccepts() {
        assertEquals(Verdict.ACCEPT, MergePolicy.decide(T, DEV_A, T, null));
    }
}

File: mobile/app/src/test/java/com/todolist/mobile/core/MergePolicyTest.kt
package com.todolist.mobile.core

import org.junit.Assert.assertEquals
import org.junit.Test

class MergePolicyTest {

    private val t = 1_700_000_000_000L // 基准 ms
    private val devA = "00000000-0000-4000-8000-00000000000a"
    private val devB = "00000000-0000-4000-8000-00000000000b"

    @Test
    fun newerClientWins() {
        assertEquals(MergePolicy.Verdict.ACCEPT, MergePolicy.decide(t + 5000, devA, t, devB))
    }

    @Test
    fun olderClientLoses() {
        assertEquals(MergePolicy.Verdict.CONFLICT, MergePolicy.decide(t - 5000, devA, t, devB))
    }

    @Test
    fun equalTimeTieBreakByDevice() {
        assertEquals(MergePolicy.Verdict.CONFLICT, MergePolicy.decide(t, devA, t, devB))
        assertEquals(MergePolicy.Verdict.ACCEPT, MergePolicy.decide(t, devB, t, devA))
    }

    @Test
    fun equalTimeSameDeviceAccepts() {
        assertEquals(MergePolicy.Verdict.ACCEPT, MergePolicy.decide(t, devA, t, devA))
    }

    @Test
    fun nullServerDeviceAccepts() {
        assertEquals(MergePolicy.Verdict.ACCEPT, MergePolicy.decide(t, devA, t, null))
    }
}
